code/datums/components/connect_range.dm

```diff
@@ -0,0 +1,91 @@
+	var/range
+	/// Whether the component works when the movable isn't directly located on a turf.
+	var/works_in_containers
+
+/datum/component/connect_range/Initialize(atom/tracked, list/connections, range, works_in_containers = TRUE)
+	if(!isatom(tracked) || isarea(tracked) || range <= 0)
+		return COMPONENT_INCOMPATIBLE
+	src.connections = connections
+	src.tracked = tracked
+	src.range = range
+	src.works_in_containers = works_in_containers
+
+/datum/component/connect_range/InheritComponent(datum/component/component, original, atom/movable/tracked, list/connections, range, works_in_containers)
+	// Not equivalent. Checks if they are not the same list via shallow comparison.
+	if(src.tracked != tracked || !compare_list(src.connections, connections))
+		return
+	if(src.range != range || src.works_in_containers != works_in_containers) //Update range and works_in_containers
+		src.range = range
+		src.works_in_containers = works_in_containers
+		update_signals(tracked, tracked.loc, TRUE)
+
+/datum/component/connect_range/RegisterWithParent()
+	RegisterSignal(tracked, COMSIG_MOVABLE_MOVED, .proc/on_moved)
+	RegisterSignal(tracked, COMSIG_PARENT_QDELETING, .proc/handle_tracked_qdel)
+	update_signals(tracked)
+
+/datum/component/connect_range/UnregisterFromParent()
+	unregister_signals(tracked)
+	UnregisterSignal(tracked, list(
+		COMSIG_MOVABLE_MOVED,
+		COMSIG_PARENT_QDELETING,
+	))
+
+	tracked = null
+
+/datum/component/connect_range/proc/handle_tracked_qdel()
+	SIGNAL_HANDLER
+	qdel(src)
+
+/datum/component/connect_range/proc/update_signals(atom/movable/movable, atom/old_loc, forced = FALSE)
+	var/turf/tracked_turf = get_turf(movable)
+	var/same_turf = !forced && tracked_turf == get_turf(old_loc) //Only register/unregister turf signals if it's moved to a new turf.
+	unregister_signals(old_loc, same_turf)
+
+	if(isnull(tracked_turf))
+		return
+
+	var/is_on_turf = tracked_turf == movable.loc
+	if(!is_on_turf && !works_in_containers) //the component doesn't register signals while inside other movables.
+		return
+
+	if(tracked_turf != movable && !is_on_turf) //the tracked atom is not a turf and is inside another movable.
+		for(var/atom/movable/container as anything in get_nested_locs(movable))
+			RegisterSignal(container, COMSIG_MOVABLE_MOVED, .proc/on_moved)
+
+	if(!same_turf)
+		for(var/turf/target_turf in RANGE_TURFS(range, tracked_turf))
+			for (var/signal in connections)
+				parent.RegisterSignal(target_turf, signal, connections[signal])
+
+/datum/component/connect_range/proc/unregister_signals(atom/location, same_turf = FALSE)
+	//The location is null or is a container and the component shouldn't have register signals on it
+	if(isnull(location) || (!works_in_containers && !isturf(location)))
+		return
+
+	if(ismovable(location))
+		UnregisterSignal(location, COMSIG_MOVABLE_MOVED)
+		for(var/atom/movable/container as anything in get_nested_locs(location))
+			UnregisterSignal(container, COMSIG_MOVABLE_MOVED)
```

Comment:
Why seperated?

Reply:
get_nested_locs doesn't include the inputted arg in its return value.